Sources/SourceKitLSP/Swift/CodeActions/PackageManifestEdits.swift

```diff
@@ -0,0 +1,254 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import PackageModel
+import PackageModelSyntax
+import SwiftRefactor
+import SwiftSyntax
+
+/// Syntactic code action provider to provide refactoring actions that
+/// edit a package manifest.
+struct PackageManifestEdits: SyntaxCodeActionProvider {
+  static func codeActions(in scope: SyntaxCodeActionScope) -> [CodeAction] {
+    guard let token = scope.firstToken,
+      let call = token.findEnclosingCall()
+    else {
+      return []
+    }
+
+    var actions = [CodeAction]()
+
+    // Add test target(s)
+    actions.append(
+      contentsOf: maybeAddTestTargetActions(call: call, in: scope)
+    )
+
+    // Add product(s).
+    actions.append(
+      contentsOf: maybeAddProductActions(call: call, in: scope)
+    )
+
+    return actions
+  }
+
+  /// Produce code actions to add test target(s) if we are currently on
+  /// a target for which we know how to create a test.
+  static func maybeAddTestTargetActions(
+    call: FunctionCallExprSyntax,
+    in scope: SyntaxCodeActionScope
+  ) -> [CodeAction] {
+    guard let calledMember = call.findMemberAccessCallee(),
+      targetsThatAllowTests.contains(calledMember),
+      let targetName = call.findStringArgument(label: "name")
+    else {
+      return []
+    }
+
+    do {
+      // Describe the target we are going to create.
+      let target = try TargetDescription(
+        name: "\(targetName)Tests",
+        dependencies: [.byName(name: targetName, condition: nil)],
+        type: .test
+      )
+
+      let edits = try AddTarget.addTarget(target, to: scope.file)
+      return [
+        CodeAction(
+          title: "Add test target",
+          kind: .refactor,
+          edit: edits.asWorkspaceEdit(snapshot: scope.snapshot)
+        )
+      ]
+    } catch {
+      return []
+    }
+  }
+
+  /// A list of target kinds that allow the creation of tests.
+  static let targetsThatAllowTests: Set<String> = [
+    "executableTarget",
+    "macro",
+    "target",
+  ]
+
+  /// Produce code actions to add a product if we are currently on
+  /// a target for which we can create a product.
+  static func maybeAddProductActions(
+    call: FunctionCallExprSyntax,
+    in scope: SyntaxCodeActionScope
+  ) -> [CodeAction] {
+    guard let calledMember = call.findMemberAccessCallee(),
+      targetsThatAllowProducts.contains(calledMember),
+      let targetName = call.findStringArgument(label: "name")
+    else {
+      return []
+    }
+
+    do {
+      let type: ProductType =
+        calledMember == "executableTarget"
+        ? .executable
+        : .library(.automatic)
+
+      // Describe the target we are going to create.
+      let product = try ProductDescription(
+        name: "\(targetName)",
```

Comment:
`targetName` is a `String`, right? So this can just be

```suggestion
        name: targetName,
```